core/src/main/scala/org/broadinstitute/dsde/rawls/entities/EntityService.scala

```diff
@@ -339,35 +344,156 @@ class EntityService(protected val ctx: RawlsRequestContext,
+        val unmarshalled = dataSource.dataAccess.entityQuery.unmarshalEntities(prevAttrs)
+        // safety check - did the attributes we gathered all marshal into a single entity?
+        if (unmarshalled.size != 1)
+          throw new DataEntityException(s"gatherOrOutput expected only one entity, found ${unmarshalled.size}")
+        AttrAccum(nextAttrs, Some(unmarshalled.head))
+      }
+
+      // inspect the variations of previous and current
+      (previous, current) match {
+        // if both previous and current are empty, it means no entities found
+        case (EmptyElement, EmptyElement) =>
+          AttrAccum(Seq(), None)
+
+        // if previous is empty but current is not, it's the first element
+        case (EmptyElement, curr: AttrAccum) =>
+          curr
+
+        // midstream, we notice that the current entity is the same as the previous entity.
+        // keep gathering attributes for this entity, and don't emit an entity yet.
+        case (prev: AttrAccum, curr: AttrAccum) if prev.accum.head.entityRecord.id == curr.accum.head.entityRecord.id =>
+          val newAccum = prev.accum ++ curr.accum
+          AttrAccum(newAccum, None)
+
+        // midstream, we notice that the current entity is DIFFERENT from the previous entity.
+        // take all the attributes we have gathered for the previous entity,
+        // marshal them into an Entity object, emit that Entity, and start a new accumulator
+        // for the new/current entity
+        case (prev: AttrAccum, curr: AttrAccum) if prev.accum.head.entityRecord.id != curr.accum.head.entityRecord.id =>
+          entityFinished(prev.accum, curr.accum)
+
+        // if current is empty but previous is not, it means the stream has finished.
+        // Marshal and output the final Entity.
+        case (prev: AttrAccum, EmptyElement) =>
+          entityFinished(prev.accum, Seq())
+
+        // relief valve, this should not happen, but if it does we should log it
+        case _ =>
+          throw new Exception(
+            s"gatherOrOutput encountered unexpected input, cannot continue. Prev: $previous :: Curr: $current"
+          )
+      }
+    }
+
+    /* custom stream stage that allows us to compare the current stream element
+       to the previous stream element. In turn, this allows us to accumulate attributes
+       until we notice that the current element is from a different entity than the previous attribute;
+       when that happens, we marshal and emit an entity.
+     */
+    class EntityCollector extends GraphStage[FlowShape[AttrAccum, AttrAccum]] {
+      val in = Inlet[AttrAccum]("EntityCollector.in")
+      val out = Outlet[AttrAccum]("EntityCollector.out")
+      override val shape = FlowShape(in, out)
+
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+        private var prev: AttributeStreamElement = EmptyElement // note: var!
```

Comment:
why the `note: var!`?

I'm curious what I should be on the lookout for here ha

Reply:
mutable variables are generally frowned upon in Scala - so this is a hint to anyone reading the code who might be assuming everything is an immutable `val`. I tried to be very careful with this mutable code - and since I can't change the signatures of the methods I'm overriding in `GraphStage` I needed to use a `var`.